[#12] 책 상세 보기 화면 설정

fried-shoes opened the pull request:
## 2025.09.08. Issue #12 에 대한 PR

### 1. 책 상세 보기 화면 추가 설정
- Toolbar 에 책 제목 표시
- 책 이미지와 내용을 본문에 배치
- API 를 통해 받는 URL 에 해당하는 책 이미지 원본이 작아서, 상세 보기 화면에서 사진이 흐리게 보이는 한계

### 2. 검색 결과 화면(`ListFragment`)에서 상세 보기 화면(`InfoFragment`) 으로 데이터 전달
- `ListFragment` 의 `Adapter` 에서 선택한 아이템의 `position` 만 받아서 `navigation argument` 로 넘김 (`Safe Args`)
- `Repository` 에 저장된 검색 결과 리스트(`searchResult`)에서 `position` 에 해당하는 책 데이터만 추출하여 표시

fried-shoes committed:
feat: 책 내용 상세 보기 화면 설정

- ViewModel 설정
- 레이아웃 설정

fried-shoes committed:
feat: ListFragment 에서 선택한 책의 데이터를 InfoFragment 로 전달

- navigation SafeArgs 를 이용하여, 선택한 아이템의 position 만 넘김
- repository 의 searchResult 에서 position 에 해당하는 데이터를 가져올 예정

fried-shoes committed:
feat: InfoFragment 에 책 상세 내용 표출

- 전달 받은 position 에 해당하는 데이터를 LiveData 에 담고, 다시 그 LiveData 를 추적하여 UI에 표출

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/InfoViewModel.kt:
```diff
@@ -0,0 +1,31 @@
+    private val _bookItem: MutableLiveData<BookItem> = MutableLiveData()
+    val bookItem: LiveData<BookItem> get() = _bookItem
+
+    fun getBookItem(itemId: Int) {
```
nit. get 이 있으면 반환을 해야하는 느낌이라서, loadBookItem 이면 어떨까 싶습니다

f-lab-tyler commented on app/src/main/res/navigation/nav_graph.xml:
```diff
@@ -26,10 +26,18 @@
         android:name="kr.yjkim.book_search.ui.InfoFragment"
-        tools:layout="@layout/fragment_info" />
+        tools:layout="@layout/fragment_info">
+        <argument
+            android:name="itemId"
+            app:argType="integer" />
```
itemId 라는 이름의 인덱스 값을 받아서 처리하는걸로 보이는데,
BookItem 객체 자체를 넘기는 방법도 있습니다.

한번 찾아서 적용해보면 ViewModel 이 좀 더 심플해지지 않을까 싶네요!

fried-shoes committed:
feat: BookItem 을 JSON string 으로 변환하여 argument 로 넘김

- moshi 객체 private 해제
- Moshi.adapter().fromJson()/toJson() 사용

fried-shoes committed:
feat: InfoViewModel 불필요하여 삭제

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/ListFragment.kt:
```diff
@@ -36,8 +38,10 @@ class ListFragment: Fragment() {
+        val adapter = BookListAdapter { bookItem ->
+            val moshiAdapter = moshi.adapter(BookItem::class.java)
+            val bookItemJson = moshiAdapter.toJson(bookItem)
+            val action = ListFragmentDirections.actionListToInfo(bookItemJson)
```
json 으로 변환하지 않아도 BookItem 이 Serializable 인터페이스를 구현하도록 하면 객체를 바로 넘길 수 있습니다!

```kotlin
@JsonClass(generateAdapter = true)
data class BookItem(
    @Json(name = "title") val title: String,
    @Json(name = "contents") val contents: String,
    @Json(name = "url") val url: String,
    @Json(name = "isbn") val isbn: String,
    @Json(name = "datetime") val datetime: String,
    @Json(name = "authors") val authors: List<String>,
    @Json(name = "publisher") val publisher: String,
    @Json(name = "translators") val translators: List<String>,
    @Json(name = "price") val price: Int,
    @Json(name = "sale_price") val salePrice: Int,
    @Json(name = "thumbnail") val thumbnail: String,
    @Json(name = "status") val status: String,
): Serializable
```

fried-shoes committed:
feat: argument 로 BookItem 을 바로 넘기도록 수정

- BookItem 이 java.io.Serializable 인터페이스를 구현하도록 함
- navigation argument 의 argType 속성 값으로 Custom Serializable class 인 BookItem 선택
